edge-modules/TwinTester/TwinOperator.cs

```diff
@@ -0,0 +1,354 @@
+// Copyright (c) Microsoft. All rights reserved.
+namespace TwinTester
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Threading.Tasks;
+    using Microsoft.Azure.Devices;
+    using Microsoft.Azure.Devices.Client;
+    using Microsoft.Azure.Devices.Edge.ModuleUtil;
+    using Microsoft.Azure.Devices.Shared;
+    using Microsoft.Extensions.Logging;
+    using Newtonsoft.Json;
+
+    public class TwinOperator
+    {
+        static readonly ILogger Logger = ModuleUtil.CreateLogger("TwinTester");
+        private readonly object operationLock = new Object();
+        private RegistryManager registryManager;
+        private ModuleClient moduleClient;
+        private AnalyzerClient analyzerClient;
+        private Storage storage;
+        private TwinState twinState;
+
+        public TwinOperator(RegistryManager registryManager, ModuleClient moduleClient, AnalyzerClient analyzerClient, Storage storage)
+        {
+            this.registryManager = registryManager;
+            this.moduleClient = moduleClient;
+            this.analyzerClient = analyzerClient;
+            this.storage = storage;
+            this.twinState = this.InitializeModuleTwin().Result;
+            this.moduleClient.SetDesiredPropertyUpdateCallbackAsync(this.OnDesiredPropertyUpdateAsync, storage);
+        }
+
+        private int GetNewPropertyCounter(Dictionary<string, DateTime> properties)
+        {
+            int maxPropertyId = -1;
+            foreach (KeyValuePair<string, DateTime> propertyUpdate in properties)
+            {
+                maxPropertyId = Math.Max(int.Parse(propertyUpdate.Key), maxPropertyId);
+            }
+
+            return maxPropertyId + 1;
+        }
+
+        private TwinCollection GetReportedPropertiesResetTwin(Twin originalTwin)
+        {
+            TwinCollection eraseReportedProperties = new TwinCollection();
+            foreach (dynamic twinUpdate in originalTwin.Properties.Reported)
+            {
+                KeyValuePair<string, object> pair = (KeyValuePair<string, object>)twinUpdate;
+                eraseReportedProperties[pair.Key] = null; // erase twin property by assigning null
+            }
+
+            return eraseReportedProperties;
+        }
+
+        private async Task<TwinState> InitializeModuleTwin()
```

Comment:
Pass cancellation token here

Reply:
No longer doing this. I'm deciding to scrap the initialization retry and just have the module fail if it cannot complete initialization. The retry policy should kick in which will restart initialization itself.